src/lib.rs

```diff
@@ -15,4 +17,22 @@ pub fn derive_key(
     let extended_public_key = extended.public_key();
     let keypair = Keypair{secret: extended.secret_key, public: extended_public_key};
     Box::new(keypair.to_bytes())
-}
\ No newline at end of file
+}
+
+/// derive_key_c does the same thing as the above function, but is intended for use over the CFFI.
+/// note that the caller must free() the returned memory as it's not managed/freed here.
+#[no_mangle]
+pub extern "C" fn derive_key_c(
+    seed: *const u8,
+    seedlen: usize,
+    path: *const u8,
+    pathlen: usize,
+) -> *mut u8 {
+    unsafe {
+        let path_str = std::str::from_utf8(std::slice::from_raw_parts(path, pathlen))
+            .expect("Failed to convert string from raw parts");
+        let seed_slice = std::slice::from_raw_parts(seed, seedlen);
+        let boxed_keypair = derive_key(seed_slice, path_str);
+        Box::into_raw(boxed_keypair) as *mut u8
```

Comment:
so how do you free it? i doubt that C.free will do the right thing. i added https://github.com/spacemeshos/go-scale/blob/1b5d749ca2366c5c5e527a09e541b8baf492437a/compat/rust/src/lib.rs#L77 for similar purpose

wasm generated it by it self:

```
export function derive_key(seed, path) {
    try {
        const retptr = wasm.__wbindgen_add_to_stack_pointer(-16);
        const ptr0 = passArray8ToWasm0(seed, wasm.__wbindgen_malloc);
        const len0 = WASM_VECTOR_LEN;
        const ptr1 = passStringToWasm0(path, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len1 = WASM_VECTOR_LEN;
        wasm.derive_key(retptr, ptr0, len0, ptr1, len1);
        var r0 = getInt32Memory0()[retptr / 4 + 0];
        var r1 = getInt32Memory0()[retptr / 4 + 1];
        var v2 = getArrayU8FromWasm0(r0, r1).slice();
        wasm.__wbindgen_free(r0, r1 * 1);
        return v2;
    } finally {
        wasm.__wbindgen_add_to_stack_pointer(16);
    }
}
```

Reply:
Sample go code is here: https://github.com/spacemeshos/smkeys/blob/f48d5bc4c5330d712bf12f8b16e0af5ac72a5458/keys.go#L39. Why do you think `C.free` won't do the right thing?

Reply:
that memory is allocated by rust and stored on rust heap, i don't understand how C.free will know how to free it. i would atleast test program heap size after running that function some vergy large number of times.

Reply:
okay, good catch, I believe you're right. added a function to clean this up correctly. some references:

- https://doc.rust-lang.org/std/ffi/struct.CString.html#method.into_raw
- https://stackoverflow.com/questions/40219711/dereference-ffi-pointer-in-python-to-get-underlying-array